src/stateMachines/services/oid4vciMachineService.ts

```diff
@@ -0,0 +1,191 @@
+  //   .then((metadata: IServerMetadataAndCryptoMatchingResponse) => {
+  //
+  //     // TODO remove brackets
+  //     return getContacts({
+  //       filter: [
+  //         {
+  //           identities: {
+  //             identifier: {
+  //               correlationId: new URL(metadata.serverMetadata.issuer).hostname,
+  //             },
+  //           },
+  //         },
+  //       ],
+  //     })
+  //   }).then((contacts: Array<IContact>): IContact | undefined => {
+  //     // TODO throw error if 0 or more than 1
+  //     return contacts.length === 1 ? contacts[0] : undefined // TODO improve code
+  //   })
+};
+
+export const retrieveCredentials = async (
+  _ctx: OID4VCIMachineContext,
+  _event: OID4VCIMachineEventTypes,
+): Promise<Array<MappedCredentialOffer> | undefined> => {
+  const {openId4VcIssuanceProvider, verificationCode, selectedCredentials} = _ctx;
+  return openId4VcIssuanceProvider
+    ?.getCredentialsFromIssuance({
+      credentials: selectedCredentials,
+      pin: verificationCode,
+    })
+    .then(
+      (credentialOffers: Array<CredentialFromOffer>): Array<MappedCredentialOffer> =>
+        credentialOffers.map((credentialOffer: CredentialFromOffer): MappedCredentialOffer => {
+          const credentialResponse: CredentialResponse = credentialOffer.credentialResponse;
+          const verifiableCredential: W3CVerifiableCredential | undefined = credentialResponse.credential;
+          const wrappedVerifiableCredential: WrappedVerifiableCredential = CredentialMapper.toWrappedVerifiableCredential(
+            verifiableCredential as OriginalVerifiableCredential,
+          );
+          const uniformVerifiableCredential: IVerifiableCredential = wrappedVerifiableCredential.credential;
+          const rawVerifiableCredential: VerifiableCredential = credentialResponse.credential as unknown as VerifiableCredential;
+          const correlationId: string =
+            typeof uniformVerifiableCredential.issuer === 'string'
+              ? uniformVerifiableCredential.issuer
+              : (uniformVerifiableCredential.issuer as IIssuer).id;
+
+          return {
+            correlationId,
+            credentialOffer,
+            rawVerifiableCredential,
+            uniformVerifiableCredential,
+          };
+        }),
+    );
+};
+
+export const addContactIdentity = async (_ctx: OID4VCIMachineContext, _event: OID4VCIMachineEventTypes): Promise<void> => {
+  const {credentialOffers, contact} = _ctx;
+  const correlationId: string = credentialOffers[0].correlationId;
+  const identity: IBasicIdentity = {
+    alias: correlationId,
+    roles: [IdentityRoleEnum.ISSUER],
+    identifier: {
+      type: CorrelationIdentifierEnum.DID,
+      correlationId,
+    },
+  };
+  store.dispatch<any>(addIdentity({contactId: contact!.id, identity}));
+};
+
+export const verifyCredentials = async (_ctx: OID4VCIMachineContext, _event: OID4VCIMachineEventTypes): Promise<void> => {
```

Comment:
Suggest to either return a boolean, or rename to assertValidCredentials

Reply:
renamed to assertValidCredentials